src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -332,115 +357,99 @@ def _post_process(
-            upload_dir,
-            counter,
-            assembly_files_len,
-        ) = task
-
-        try:
-            upa = _upload_assembly_to_workspace(
-                conf, workspace_id, container_internal_assembly_path, assembly_name
-            )
-            _post_process(
-                upload_env_key,
-                workspace_id,
-                host_assembly_dir,
-                assembly_name,
-                upload_dir,
-                conf.output_dir,
-                upa
-            )
-        except Exception as e:
-            print(f"Failed assembly name: {assembly_name}. Exception:")
-            print(e)
-
-        conf.output_queue.put((assembly_name, upa))
-
-        if counter % 3000 == 0:
-            print(f"Assemblies processed: {counter}/{assembly_files_len}, "
-                  f"Percentage: {counter / assembly_files_len * 100:.2f}%, "
-                  f"Time: {datetime.now()}")
-
-
 def _upload_assembly_files_in_parallel(
-        conf: Conf,
+        asu: AssemblyUtil,
+        ws: Workspace,
         upload_env_key: str,
         workspace_id: int,
         upload_dir: str,
         wait_to_upload_assemblies: dict[str, str],
+        batch_size: int,
+        output_dir: str,
 ) -> list[str]:
     """
     Upload assembly files to the target workspace in parallel using multiprocessing.
 
     Parameters:
-        conf: Conf object
+        asu: AssemblyUtil client
+        ws: Workspace client
         upload_env_key: environment variable key in uploaded.yaml file
         workspace_id: target workspace id
         upload_dir: a directory in collectionssource that creates new directories linking to sourcedata
         wait_to_upload_assemblies: a dictionary that maps assembly file name to assembly directory
+        batch_size: a number of files to upload per batch
+        output_dir: a directory in sourcedata/workspace to store new assembly entries
 
     Returns:
-        a list of assembly names that failed to upload
+        number of assembly files have been sucessfully uploaded from wait_to_upload_assemblies
     """
     assembly_files_len = len(wait_to_upload_assemblies)
-    print(f"Start uploading {assembly_files_len} assembly files with {conf.workers} workers\n")
+    print(f"Start uploading {assembly_files_len} assembly files\n")
 
-    # Put the assembly files in the input_queue
-    counter = 1
-    for assembly_name, host_assembly_dir in wait_to_upload_assemblies.items():
+    uploaded_count = 0
+    uploaded_fail = False
+    for assembly_tuples in _gen(wait_to_upload_assemblies, batch_size):
 
-        container_internal_assembly_path = os.path.join(
-            JOB_DIR_IN_ASSEMBLYUTIL_CONTAINER, assembly_name
-        )
-        conf.input_queue.put(
-            (
+        try:
+            batch_upas = _upload_assemblies_to_workspace(asu, workspace_id, assembly_tuples)
+            batch_uploaded_tuples = assembly_tuples
+        except Exception as e:
```

Comment:
It looks like this exception is never even printed out, which makes debugging difficult

Reply:
the exception has been printed out at least 2 times. so I skip it here

Reply:
Maybe it should be printed here instead of earlier since this is where the exception bubbling terminates

Reply:
Updated!

Reply:
There's no need for the try catch in _uatw any more - this is effectively a noop:
```python
    except Exception as e:
        raise e
```

Reply:
👍